---
title: "Importance markers"
author: "Costa, W. G."
date: "`r Sys.Date()`"
site: workflowr::wflow_site
url: https://wevertongomescosta.github.io/Importance-of-markers-for-QTL-detection-by-machine-learning-methods/
output:
  workflowr::wflow_html:
    toc: TRUE
  highlight: github
editor_options:
  chunk_output_type: console
github-repo: wevertongomescosta/Importance-of-markers-for-QTL-detection-by-machine-learning-methods
---

# Data

## Data importantion

The traits importance data was obtained by script in [Genomic-prediction-through-machine-learning-and-neural-networks-for-traits-with-epistasis](https://wevertongomescosta.github.io/Genomic-prediction-through-machine-learning-and-neural-networks-for-traits-with-epistasis/)

# Packages
```{r}
library(metan)
library(tidyverse)
library(ggrepel)
library(ggnewscale)
library(ggthemes)
library(gridExtra)
```

# Carregando os dados
```{r}
load("input/imp.tot.RData")
head(imp.tot)
load("input/snpsOfInterest.RData")
head(snpsOfInterest)
load("input/results_gwas.rda")
head(results_gwas)
```

# Preparando os dados de importância

```{r}
imp.tot <- imp.tot %>%
  mutate(
    variable = as.numeric(variable),
    method = as.factor(method),
    marker = as.numeric(str_replace(marker, "V", ""))
  ) %>%
  group_by(variable, method, marker) %>%
  summarise(Overall = mean(Overall), .groups = "drop")
```

# Preparando os resultados do GWAS

```{r}
# Definindo o limiar de significância
alpha <- 0.05
n.markers <- 4010
threshold <- alpha / n.markers

# Preparando os resultados do GWAS
results_gwas <- results_gwas %>%
  rename(marker = SNP, CHR = Chr) %>%
  mutate(
    variable = as.numeric(str_replace(variable, "V", "")),
    method = as.factor(method),
    CHR = as.numeric(CHR),
    marker = as.numeric(marker),
    Overall = -log10(P.value)
  )


results_gwas_sig <- results_gwas %>%
  filter(P.value < threshold) %>%
  mutate(sig = paste(variable, marker, sep = "_")) %>%
  droplevels() %>%
  select(variable, method, marker, CHR, Overall, sig)
```

# Normalizando a importância dos marcadores para cada variável e método

```{r}
methods <- unique(imp.tot$method)
variables <- unique(imp.tot$variable)

normalized_imp <- imp.tot %>%
  group_by(method, variable) %>%
  mutate(Overall_res = scales::rescale(Overall, to = c(0, 10), finite = F)) %>%
  ungroup()
```


# Primeiro, atribuímos 'region_markers' dentro de cada grupo de 'variable'

```{r}
reg1 <- snpsOfInterest %>%
  group_by(variable) %>%
  mutate(region_markers = row_number()) %>%
  ungroup()

# Agora, criamos o data frame 'region' com os marcadores ao redor de cada SNP de interesse
distance <- 5


region <- reg1 %>%
  rowwise() %>%
  mutate(marker = list(seq(marker - distance, marker + distance))) %>%
  ungroup() %>%
  unnest(cols = marker) %>%
  filter(variable %in% c(1:5, 9:13)) %>%
  droplevels()
```

# Contando o número de genes em cada região

```{r}
ngenes_region <- region %>%
  group_by(variable) %>%
  summarise(ngenes_region = n(), .groups = "drop") %>%
  mutate(variable = factor(variable, labels = c(1:10)))
```

# Calculando a importância média dos marcadores

```{r}
effects_markers <- normalized_imp %>%
  group_by(variable, method) %>%
  summarise(effects_markers = mean(Overall_res), .groups = "drop")
```

# Associando os marcadores às regiões do QTL

```{r}
plot_data <- normalized_imp %>%
  mutate(
    CHR = case_when(
      marker <= 401 ~ 1,
      marker <= 802 ~ 2,
      marker <= 1203 ~ 3,
      marker <= 1604 ~ 4,
      marker <= 2005 ~ 5,
      marker <= 2406 ~ 6,
      marker <= 2807 ~ 7,
      marker <= 3208 ~ 8,
      marker <= 3609 ~ 9,
      TRUE ~ 10
    )
  ) %>%
  full_join(results_gwas) %>%
  full_join(effects_markers) %>%
  full_join(region)

plot_data <- plot_data %>%
  mutate(
    sig = paste(variable, marker, sep = "_"),
    is_highlight = if_else(!is.na(region_markers), "yes", "no"),
    is_annotate = case_when(
      Overall_res > effects_markers & method != "GLM" ~ "yes",
      method == "GLM" & sig %in% results_gwas_sig$sig ~ "yes",
      TRUE ~ "no",
    )
  )
```

# Criando variáveis adicionais

```{r}
plot_data <- plot_data %>%
  mutate(
    herd = factor(if_else(variable < 9, "50%", "80%")),
    ngenes = factor(
      case_when(
        variable %in% c(1, 9) ~ 8,
        variable %in% c(2, 10) ~ 40,
        variable %in% c(3, 11) ~ 80,
        variable %in% c(4, 12) ~ 120,
        variable %in% c(5, 13) ~ 240,
        variable %in% c(6, 14) ~ 480,
        variable %in% c(7, 15) ~ 88,
        variable %in% c(8, 16) ~ 160
      )
    ),
    created = factor(if_else(
      ngenes %in% c(88, 160), "Simulated modified", "Simulated"
    )),
    is_high_anno = case_when(
      is_highlight == "yes" & is_annotate == "yes" ~ 1,
      is_highlight == "yes" & is_annotate == "no" ~ 2,
      is_highlight == "no" & is_annotate == "yes" ~ 3,
      TRUE ~ 4
    ),
    variable = factor(variable)
  )


levels(plot_data$method) <-  c("BAG",
                               "BO",
                               "DT",
                               "G-BLUP",
                               "MARS 3",
                               "MARS 1",
                               "MARS 2",
                               "MLP",
                               "RBF",
                               "RF",
                               "GLM")

plot_data <- plot_data %>%
  mutate(method =  fct_relevel(method, sort))
```

# Filtrando apenas as variáveis simuladas

```{r}
plot_data <- plot_data %>%
  filter(created == "Simulated" & ngenes != 480) %>%
  #filter(!(method %in% c("RBF", "MLP", "G-BLUP"))) %>%
  droplevels()

levels(plot_data$variable) <-
  c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")
```

# Calculando o índice de coincidência

```{r}
ver <- plot_data %>%
  filter(is_annotate == "yes") %>%
  group_by(variable, method, CHR, region_markers) %>%
  summarise(n = n(), .groups = "drop")
```

# Separando acertos e erros

```{r}
certo <- ver %>%
  filter(region_markers != "NA") %>%
  group_by(variable, method) %>%
  reframe(n = n)

acerto <- certo %>%
  count(variable, method, name = "acerto")

errado <- ver %>%
  filter(is.na(region_markers)) %>%
  group_by(variable, method, CHR, region_markers) %>%
  reframe("erro" = n)

erro <- errado %>%
  group_by(variable, method) %>%
  summarise(erro = sum(erro), .groups = "drop")
```

# Unindo acertos e erros e calculando os índices

```{r}
coinc <- full_join(acerto, erro) %>%
  full_join(ngenes_region) %>%
  mutate(
    acerto = replace_na(acerto, 0),
    erro = replace_na(erro, 0),
    ngenes = case_when(
      variable %in% c("1", "6") ~ 8,
      variable %in% c("2", "7") ~ 40,
      variable %in% c("3", "8") ~ 80,
      variable %in% c("4", "9") ~ 120,
      variable %in% c("5", "10") ~ 240
    ),
    herd = if_else(as.numeric(variable) < 6, "h² = 0.50", "h² = 0.80"),
    # Poder de Decisão (PD) = Sensibilidade
    PD = round((acerto / ngenes) * 100, 2),
    # Falso Positivo (FP)
    FP = round((erro / (
      n.markers - ngenes_region
    )) * 100, 2),
    # Precisão
    Precision = round(acerto / (acerto + erro) * 100, 2),
    # Adicionando os novos índices:
    Specificity = round((((n.markers - ngenes_region) - erro
    ) / (
      n.markers - ngenes_region
    )) * 100, 2),
    # Média harmônica entre precisão e Poder de Decisão ,
    `F1 Score` = round(2 * (Precision * PD) / (Precision + PD), 2),
  ) %>%
  filter(!(method %in% c("RBF", "MLP", "G-BLUP"))) %>%
  droplevels()

acerto_erro <- coinc %>%
  select(variable, method, acerto, erro)

coinc %>%
  select(variable, method, PD) %>%
  pivot_wider(names_from = variable, values_from = PD)


writexl::write_xlsx(acerto_erro, "output/acerto_erro.xlsx")
```

# Criando o gráfico

## PD, Precision
```{r}
coinc %>%
  select(method, ngenes, herd, PD, Precision) %>%
  pivot_longer(cols = 4:5) %>%
  mutate(method = factor(method, levels = rev(sort(unique(
    method
  ))))) %>%
  ggplot(aes(
    x = factor(ngenes),
    y = method,
    fill = value
  )) +
  
  # Adicionando os tiles (blocos da matriz)
  geom_tile(color = "black") +
  
  # Ajustando a escala de preenchimento
  scale_fill_gradient2(
    low = "white",
    mid = "yellow",
    high = "red",
    midpoint = 50,
    # 100 / 2 simplifica para 50
    limits = c(0, 100),
    space = "Lab",
    name = "Value"
  ) +
  
  # Facetando por 'herd' (herdabilidade)
  facet_wrap(name ~ herd, nrow = 1) +
  
  # Aplicando tema minimalista com tamanho de texto base
  theme_bw() +
  
  # Ajustando elementos do tema
  theme(
    axis.text.x = element_text(vjust = 1, hjust = 0.5),
    axis.title.y = element_blank(),
    panel.grid.major = element_blank(),
    axis.ticks = element_blank(),
    legend.position = "top",
    legend.direction = "horizontal",
    #    legend.title = element_text(size = 12),
    #    legend.text = element_text(size = 10)
  ) +
  
  # Fixando a proporção dos eixos
  coord_fixed() +
  
  # Adicionando os valores de PD como texto nos tiles
  geom_text(aes(label = value), color = "black") +
  
  # Ajustando o título do eixo X
  labs(x = "Number of Genes")
```

## FP, Specificity

```{r}
coinc %>%
  select(method, ngenes, herd, FP, Specificity) %>%
  pivot_longer(cols = 4:5) %>%
  mutate(method = factor(method, levels = rev(sort(unique(
    method
  ))))) %>%
  ggplot(aes(
    x = factor(ngenes),
    y = method,
    fill = value
  )) +
  
  # Adicionando os tiles (blocos da matriz)
  geom_tile(color = "black") +
  
  # Ajustando a escala de preenchimento
  scale_fill_gradient2(
    low = "white",
    mid = "yellow",
    high = "red",
    midpoint = 50,
    # 100 / 2 simplifica para 50
    limits = c(0, 100),
    space = "Lab",
    name = "Value"
  ) +
  
  # Facetando por 'herd' (herdabilidade)
  facet_wrap(name ~ herd, nrow = 1) +
  
  # Aplicando tema minimalista com tamanho de texto base
  theme_bw() +
  
  # Ajustando elementos do tema
  theme(
    axis.text.x = element_text(vjust = 1, hjust = 0.5),
    axis.title.y = element_blank(),
    panel.grid.major = element_blank(),
    axis.ticks = element_blank(),
    legend.position = "top",
    legend.direction = "horizontal",
    #    legend.title = element_text(size = 12),
    #    legend.text = element_text(size = 10)
  ) +
  
  # Fixando a proporção dos eixos
  coord_fixed() +
  
  # Adicionando os valores de PD como texto nos tiles
  geom_text(aes(label = value), color = "black") +
  
  # Ajustando o título do eixo X
  labs(x = "Number of Genes")
```

## F1 Score

```{r}
coinc %>%
  select(method, ngenes, herd, `F1 Score`) %>%
  pivot_longer(cols = 4) %>%
  mutate(method = factor(method, levels = rev(sort(unique(
    method
  ))))) %>%
  ggplot(aes(
    x = factor(ngenes),
    y = method,
    fill = value
  )) +
  
  # Adicionando os tiles (blocos da matriz)
  geom_tile(color = "black") +
  
  # Ajustando a escala de preenchimento
  scale_fill_gradient2(
    low = "white",
    mid = "yellow",
    high = "red",
    midpoint = 50,
    # 100 / 2 simplifica para 50
    limits = c(0, 100),
    space = "Lab",
    name = "Value"
  ) +
  
  # Facetando por 'herd' (herdabilidade)
  facet_wrap(name ~ herd, nrow = 1) +
  
  # Aplicando tema minimalista com tamanho de texto base
  theme_bw() +
  
  # Ajustando elementos do tema
  theme(
    axis.text.x = element_text(vjust = 1, hjust = 0.5),
    axis.title.y = element_blank(),
    panel.grid.major = element_blank(),
    axis.ticks = element_blank(),
    legend.position = "right",
    legend.direction = "vertical",
    #    legend.title = element_text(size = 12),
    #    legend.text = element_text(size = 10)
  ) +
  
  # Fixando a proporção dos eixos
  coord_fixed() +
  
  # Adicionando os valores de PD como texto nos tiles
  geom_text(aes(label = value), color = "black") +
  
  # Ajustando o título do eixo X
  labs(x = "Number of Genes")
```

# Manhattan plot

```{r}
###### Definindo as posi??es dos grupos de liga??o no plot
axisdf = plot_data %>%
  group_by(CHR) %>%
  summarize(center = (max(marker) + min(marker)) / 2)
```


```{r}
for (i in 1:10) {
  plotgwasall <- plot_data %>%
    filter(variable == i)
  
  region <- plotgwasall %>%
    filter(variable == i, region_markers > 0) %>%
    group_by(region_markers) %>%
    summarise(marker_min = min(marker),
              marker_max = max(marker))
  
  p <- ggplot() +
    geom_point(
      data = plotgwasall,
      aes(marker, Overall_res, color = as.factor(CHR)),
      alpha = 0.05,
      size = 1
    ) +
    scale_x_continuous(label = axisdf$CHR, breaks = axisdf$center) +
    #scale_y_continuous(expand = c(0.01, 0.05), breaks = c(0, 100)) + #remove space between plot area and x axis
    geom_hline(yintercept = 1.25,
               linetype = "dashed",
               color = "black") +
    # Add highlighted points
    geom_rect(
      data = region,
      aes(xmin = marker_min, xmax = marker_max),
      ymin = Inf,
      ymax = Inf,
      alpha = 0.25,
      fill = "orange"
    ) +
    facet_wrap(
      method ~ .,
      ncol = 1,
      scales = "free_y" ,
      strip.position = "right"
    ) +
    theme_hc() +
    scale_color_gdocs() +
    new_scale_color() +
    geom_point(
      data = subset(plotgwasall, is_high_anno != 4),
      aes(marker, Overall_res, color = as.factor(is_high_anno)),
      size = 2
    ) +
    scale_color_manual(values = c("seagreen3", "orange", "red")) +
    xlab("Linkage Group") + ylab("Importance (%)") +
    theme(
      legend.position = "none",
      text = element_text(size = 12),
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      panel.spacing.y = unit(1, "lines"),
      axis.ticks.x = element_blank()
    )
  
  print("Trait:", i)
  print(p)
}
```
