---
title: "GWAS"
author: "Costa, W. G."
date: "`r Sys.Date()`"
site: workflowr::wflow_site
url: https://wevertongomescosta.github.io/Importance-of-markers-for-QTL-detection-by-machine-learning-methods/
output:
  workflowr::wflow_html:
    toc: yes
  highlight: github
editor_options:
  chunk_output_type: console
github-repo: wevertongomescosta/Importance-of-markers-for-QTL-detection-by-machine-learning-methods
---

## Importance of markers for QTL detection by machine learning methods

# Creating Genetic Map Chart

## Libraries 

```{r}
library(tidyverse)
library(data.table)
library(metan)
library(ggthemes)
library(ggrepel)
library(ggpubr)
library(cowplot)
library(tidytext)
```

## Phenotypic variance and genetic values

```{r}
pheno <- read.table("data/simulated_data/phenotype.txt")

colnames(pheno) <- str_replace_all(colnames(pheno), "V", "P")

genetic_values <-
  read.table("data/simulated_data/genotypic values.txt")

colnames(genetic_values) <-
  str_replace_all(colnames(genetic_values), "V", "G")

pheno <- pheno %>%
  pivot_longer(1:18) %>%
  mutate(trait = as.numeric(factor(name)))

genetic_values <-  genetic_values %>%
  pivot_longer(1:18) %>%
  mutate(trait = as.numeric(factor(name)))

data <- pheno %>%
  full_join(genetic_values) %>%
  mutate(type = ifelse(name %like% "P", "Phenotype", "Genetic"))

data  %>%
  ggplot(aes(x = as.factor(trait), y = value, fill = type)) +
  geom_boxplot() +
  ylim(0, 120) +
  scale_fill_gdocs() +
  theme_bw() +
  theme(text = element_text(size = 20, face = "bold")) +
  labs(x = "Traits",
       y = "Value",
       fill = "Type")
```

## Genetic map

Primeiro vamos definir nossos SNPs de interesse para as variáveis. Essas informações foram pré-definidas e podem ser encontrados no arquivo `control_genetic`![](data/simulated_data/control_genetic.txt). Vamos carregar o arquivo `snpsOfInterest.RData`"[](analysis/snpsOfInterest.RData) que comtemas infomações dos snps considerados QTLs por variavel. 

```{r}
load("data/snpsOfInterest.RData")
```

Agora vamos definir os nomes das colunas para nosso `snpsOfInterest` e criar um objeto `locus` com as informações de ininício e témino de cada grupo de ligação. 

```{r}
locus <-
  data.frame(
    c(
      1,
      401,
      402,
      802,
      803,
      1203,
      1204,
      1604,
      1605,
      2005,
      2006,
      2406,
      2407,
      2807,
      2808,
      3208,
      3209,
      3609,
      3610,
      4010
    )
  )

colnames(locus) <- c("marker")
```

Para facilitar a visualização criei um gráfico  genético das características `map_plot`. Para isso criei o  `map` com o número de marcadores e tamanho de cada grupo de liagação. 

```{r}
map <-
  data.frame(rbind(
    cbind(seq(1, 401, 1), rep("LG 1", 401), seq(0, 200, 0.5)),
    cbind(seq(402, 802, 1), rep("LG 2", 401), seq(0, 200, 0.5)),
    cbind(seq(803, 1203, 1), rep("LG 3", 401), seq(0, 200, 0.5)),
    cbind(seq(1204, 1604, 1), rep("LG 4", 401), seq(0, 200, 0.5)),
    cbind(seq(1605, 2005, 1), rep("LG 5", 401), seq(0, 200, 0.5)),
    cbind(seq(2006, 2406, 1), rep("LG 6", 401), seq(0, 200, 0.5)),
    cbind(seq(2407, 2807, 1), rep("LG 7", 401), seq(0, 200, 0.5)),
    cbind(seq(2808, 3208, 1), rep("LG 8", 401), seq(0, 200, 0.5)),
    cbind(seq(3209, 3609, 1), rep("LG 9", 401), seq(0, 200, 0.5)),
    cbind(seq(3610, 4010, 1), rep("LG 10", 401), seq(0, 200, 0.5))
  ))

colnames(map) <- c("marker", "LG", "Size")

map <- map %>%
  mutate(
    marker = as.numeric(marker),
    Size = as.numeric(Size),
    LG = factor(
      LG,
      levels = c(
        "LG 1",
        "LG 2",
        "LG 3",
        "LG 4",
        "LG 5",
        "LG 6",
        "LG 7",
        "LG 8",
        "LG 9",
        "LG 10"
      )
    )
  )
```

Para dividir a figura e mostrar todos os maps genômicos das características, dividi o `snpsOfInterest` e o `map` para cada característica e inclui os SNPs de interesse no `map` para cada característica.

```{r}
snpsOfInterest1 <- snpsOfInterest %>%
  filter(variable == 1)

snpsOfInterest2 <- snpsOfInterest %>%
  filter(variable == 2)

snpsOfInterest3 <- snpsOfInterest %>%
  filter(variable == 3)

snpsOfInterest4 <- snpsOfInterest %>%
  filter(variable == 4)

snpsOfInterest5 <- snpsOfInterest %>%
  filter(variable == 5)

map1 <- map %>%
  mutate(
    is_highlight = ifelse(marker %in% snpsOfInterest1$marker, "yes", "no"),
    is_locus = ifelse(marker %in% locus$marker, "yes", "no")
  )

map2 <- map %>%
  mutate(
    is_highlight = ifelse(marker %in% snpsOfInterest2$marker, "yes", "no"),
    is_locus = ifelse(marker %in% locus$marker, "yes", "no")
  )

map3 <- map %>%
  mutate(
    is_highlight = ifelse(marker %in% snpsOfInterest3$marker, "yes", "no"),
    is_locus = ifelse(marker %in% locus$marker, "yes", "no")
  )

map4 <- map %>%
  mutate(
    is_highlight = ifelse(marker %in% snpsOfInterest4$marker, "yes", "no"),
    is_locus = ifelse(marker %in% locus$marker, "yes", "no")
  )

map5 <- map %>%
  mutate(
    is_highlight = ifelse(marker %in% snpsOfInterest5$marker, "yes", "no"),
    is_locus = ifelse(marker %in% locus$marker, "yes", "no")
  )
```

Agora cirei o gráfico de cada característica e depois agrupei eles em apenas uma imagem `maps`.

```{r}
map_plot1 <- ggplot(map1, aes(x = LG, y = Size)) +
  geom_segment(aes(
    yend = 200,
    y = 0,
    x = LG,
    xend = LG
  ),
  color = "skyblue",
  size = 1) +
  geom_point(
    data = subset(map1, is_locus == "yes"),
    color = "skyblue",
    size = 0.5
  ) +
  geom_point(
    data = subset(map1, is_highlight == "yes"),
    color = "Orange",
    size = 0.5
  ) +
  geom_text_repel(
    data = subset(map1, is_highlight == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  geom_text_repel(
    data = subset(map1, is_locus == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  scale_x_discrete(expand = expansion(mult = c(0.15, 0.05))) +
  scale_y_continuous(expand = expansion(mult = c(0.03, 0.05))) +
  theme_void() +
  theme(
    axis.text.y = element_blank(),
    axis.text.x = element_text(size = 4),
    axis.ticks = element_blank()
  ) +
  labs(y = "", x = "")

map_plot2 <- ggplot(map2, aes(x = LG, y = Size)) +
  geom_segment(aes(
    yend = 200,
    y = 0,
    x = LG,
    xend = LG
  ),
  color = "skyblue",
  size = 1) +
  geom_point(
    data = subset(map2, is_locus == "yes"),
    color = "skyblue",
    size = 0.5
  ) +
  geom_point(
    data = subset(map2, is_highlight == "yes"),
    color = "Orange",
    size = 0.5
  ) +
  geom_text_repel(
    data = subset(map2, is_highlight == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  geom_text_repel(
    data = subset(map2, is_locus == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  scale_x_discrete(expand = expansion(mult = c(0.15, 0.05))) +
  scale_y_continuous(expand = expansion(mult = c(0.03, 0.05))) +
  theme_void() +
  theme(
    axis.text.y = element_blank(),
    axis.text.x = element_text(size = 4),
    axis.ticks = element_blank()
  ) +
  labs(y = "", x = "")

map_plot3 <- ggplot(map3, aes(x = LG, y = Size)) +
  geom_segment(aes(
    yend = 200,
    y = 0,
    x = LG,
    xend = LG
  ),
  color = "skyblue",
  size = 1) +
  geom_point(
    data = subset(map3, is_locus == "yes"),
    color = "skyblue",
    size = 0.5
  ) +
  geom_point(
    data = subset(map3, is_highlight == "yes"),
    color = "Orange",
    size = 0.5
  ) +
  geom_text_repel(
    data = subset(map3, is_highlight == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  geom_text_repel(
    data = subset(map3, is_locus == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  scale_x_discrete(expand = expansion(mult = c(0.15, 0.05))) +
  scale_y_continuous(expand = expansion(mult = c(0.03, 0.05))) +
  theme_void() +
  theme(
    axis.text.y = element_blank(),
    axis.text.x = element_text(size = 4),
    axis.ticks = element_blank()
  ) +
  labs(y = "", x = "")

map_plot4 <- ggplot(map4, aes(x = LG, y = Size)) +
  geom_segment(aes(
    yend = 200,
    y = 0,
    x = LG,
    xend = LG
  ),
  color = "skyblue",
  size = 1) +
  geom_point(
    data = subset(map4, is_locus == "yes"),
    color = "skyblue",
    size = 0.5
  ) +
  geom_point(
    data = subset(map4, is_highlight == "yes"),
    color = "Orange",
    size = 0.5
  ) +
  geom_text_repel(
    data = subset(map4, is_highlight == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  geom_text_repel(
    data = subset(map4, is_locus == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  scale_x_discrete(expand = expansion(mult = c(0.15, 0.05))) +
  scale_y_continuous(expand = expansion(mult = c(0.03, 0.05))) +
  theme_void() +
  theme(
    axis.text.y = element_blank(),
    axis.text.x = element_text(size = 4),
    axis.ticks = element_blank()
  ) +
  labs(y = "", x = "")

map_plot5 <- ggplot(map5, aes(x = LG, y = Size)) +
  geom_segment(aes(
    yend = 200,
    y = 0,
    x = LG,
    xend = LG
  ),
  color = "skyblue",
  size = 1) +
  geom_point(
    data = subset(map5, is_locus == "yes"),
    color = "skyblue",
    size = 0.5
  ) +
  geom_point(
    data = subset(map5, is_highlight == "yes"),
    color = "Orange",
    size = 0.5
  ) +
  geom_text_repel(
    data = subset(map5, is_highlight == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  geom_text_repel(
    data = subset(map5, is_locus == "yes"),
    aes(label = marker),
    size = 1.5,
    max.overlaps = Inf,
    min.segment.length = 0,
    force   = 0,
    nudge_x      = -0.55,
    nudge_y      = -1.5,
    direction    = "x",
    hjust        = 0.5,
    segment.curvature = -1e-20,
    segment.angle     = 45,
    segment.size = 0.1
  ) +
  scale_x_discrete(expand = expansion(mult = c(0.15, 0.05))) +
  scale_y_continuous(expand = expansion(mult = c(0.03, 0.05))) +
  theme_void() +
  theme(
    axis.text.y = element_blank(),
    axis.text.x = element_text(size = 4),
    axis.ticks = element_blank()
  ) +
  labs(y = "", x = "")

maps <- ggdraw() +
  draw_plot(
    map_plot1,
    x = 0.05,
    y = .5,
    width = .3,
    height = .5
  ) +
  draw_plot(
    map_plot2,
    x = .4,
    y = .5,
    width = .3,
    height = .5
  ) +
  draw_plot(
    map_plot3,
    x = .75,
    y = .5,
    width = .3,
    height = .5
  ) +
  draw_plot(
    map_plot4,
    x = 0.25,
    y = 0,
    width = 0.3,
    height = 0.5
  ) +
  draw_plot(
    map_plot5,
    x = 0.65,
    y = 0,
    width = 0.3,
    height = 0.5
  ) +
  draw_plot_label(
    label = c("A", "B", "C", "D", "E"),
    size = 15,
    x = c(0, 0.35, 0.7, 0.2, 0.6),
    y = c(1, 1, 1, 0.5, 0.5)
  )

print(maps)
```


# Data

## Data importantion

The traits importance data was obtained by script in [Genomic-prediction-through-machine-learning-and-neural-networks-for-traits-with-epistasis](https://wevertongomescosta.github.io/Genomic-prediction-through-machine-learning-and-neural-networks-for-traits-with-epistasis/)

```{r}
imp <- list.files(path = "data/imp/",
                  pattern = "imp.",
                  full.names = TRUE) %>%
  map_df(~ get(load(file = .x)))
```

##### Normalizando a importa?cia dos marcadores para cada variavel de cada metodo #######

Pra que as metodologias possam ser comparadas nós normalizamos os dados na escala de 0 a 100.

```{r}
for (j in levels(imp$method)) {
  for (i in 1:16) {
    x <- imp %>%
      filter (method == j & variable == i)
    
    x <- resca(x, Overall, new_min = 0, new_max = 100)
    
    if (i == 1) {
      x1 <- x
    } else {
      x1 <- x1 %>% rbind(x)
    }
  }
  
  if (j == levels(imp$method)[1]) {
    x2 <- x1
  } else {
    x2 <- x2 %>% rbind(x1)
  }
  
}
imp <- x2

rm(x,x1,x2,i,j)

```

##### Obtendo a média de cada marcador entre os folds

```{r}
imp <- imp %>%
  group_by(variable, method, marker) %>%
  summarise(Overall = mean(Overall))
```

###### Criando uma coluna referente aos grupos de liga??o #####


```{r}
imp <- imp %>%
  mutate(marker = as.numeric(marker),
    CHR = case_when(
      marker <= 401 ~ 1,
      marker > 401 & marker <= 802 ~ 2,
      marker > 802 & marker <= 1203 ~ 3,
      marker > 1203 & marker <= 1604 ~ 4,
      marker > 1604 & marker <= 2005 ~ 5,
      marker > 2005 & marker <= 2406 ~ 6,
      marker > 2406 & marker <= 2807 ~ 7,
      marker > 2807 & marker <= 3208 ~ 8,
      marker > 3208 & marker <= 3609 ~ 9,
      marker > 3609 ~ 10
    )
  )
```



# Criando a coluna representando a região do QTL


```{r}

for(i in 1:16) {
  reg <- snpsOfInterest %>%
    filter(variable == i)
  for (j in 1:nrow(reg)) {
    reg$region_markers[j] <- j
  }
  if (i == 1) {
    reg1 <- reg
  }
  if (i != 1) {
    reg1 <- rbind(reg1, reg)
  }
}


for (i in 1:nrow(snpsOfInterest)) {
  marker <- c(
    seq.int(snpsOfInterest$marker[i] - 4, snpsOfInterest$marker[i]),
    seq.int(snpsOfInterest$marker[i] + 1, snpsOfInterest$marker[i] + 4)
  )
  variable <- snpsOfInterest$variable[i]
  region_markers <- reg1$region_markers[i]
  if (i == 1) {
    region <- as.data.frame(cbind(variable, marker, region_markers))
  } else {
    region1 <- as.data.frame(cbind(variable, marker, region_markers))
    region <- rbind(region, region1)
  }
}
```

##### Criando as colunas de highlight e annotate dos marcadores de interesse e mais importantes ####

```{r}
for (j in levels(imp$variable)) {
  snpsOfInterest1 = region %>%
    filter(variable == j)
  imp1 = imp %>%
    filter(variable == j)
  imp1 <- imp1 %>%
    mutate(is_highlight = ifelse(marker %in% snpsOfInterest1$marker, "yes", "no"))
  if (j == 1) {
    imp2 = imp1
  } else {
    imp2 = rbind(imp2, imp1)
  }
}

imp <- imp2

imp <- imp %>%
  mutate(is_annotate = ifelse(Overall > 12.5, "yes", "no"))
```

###### Definindo as posi??es dos grupos de liga??o no plot

```{r}
axisdf <- imp %>% group_by(CHR) %>% summarize(center = (max(marker) + min(marker)) /2)
```

####### Gr?ficos

```{r}
for (i in levels(imp$method)) {
  dataplot <- imp %>%
    filter(method == i) %>% 
    droplevels()

  #Show all points
  dataplot <- dataplot %>%
    ggplot(aes(marker, Overall)) +
    facet_wrap( ~ variable, ncol = 3) +
    geom_point(aes(color = as.factor(CHR)), alpha = 0.8, size = 1.3) +
    scale_color_manual(values = rep(c("grey", "skyblue"), 22)) +

    #custom X axis:
    scale_x_continuous(label = axisdf$CHR, breaks = axisdf$center) +
    scale_y_continuous(expand = c(0.01, 0.05)) + #remove space between plot area and x axis
    geom_hline(yintercept = 50,
               linetype = "dashed",
               color = "red") +

    # Add highlighted points
    geom_point(
      data = subset(dataplot, is_annotate == "yes"),
      color = "Orange",
      size = 2,
      box.padding = 0.5
    ) +
    geom_point(
      data = subset(dataplot, is_highlight == "yes"),
      color = "seagreen3",
      size = 2,
      box.padding = 0.5
    ) +

    # Add label using ggrepel to avoid overlapping
    geom_label_repel(
      data = subset(dataplot, is_annotate == "yes"),
      aes(label = marker),
      size = 2,
      color = "Orange",
      box.padding = 0.5
    ) +
    geom_label_repel(
      data = subset(dataplot, is_highlight == "yes"),
      aes(label = marker),
      size = 2,
      color = "seagreen3",
      box.padding = 0.5
    ) +

    #Custom the theme:
    theme_bw() +
    theme(
      text = element_text(size = 30),
      legend.position = "none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) +
    labs(x = "Grupos de Liga??o",
         y = "Import?ncia")
  print(dataplot)
  
  nome <- paste("output/GWAS", i, ".tiff", sep = "")
  ggsave(
    nome,
    plot = dataplot,
    width = 24,
    height = 16,
    dpi = 300
  )
}
```

# Indice de coincidência

```{r}
ver <- imp %>%
  filter(is_annotate == "yes")%>%
  group_by(variable, method, CHR, region_markers)%>%
  summarise(n = n())  %>%
  ungroup()

levels(ver$method)

certo<- ver %>%
  filter(region_markers != "NA") %>%
  group_by(variable, method) %>%
  summarise(n = n)%>%
  ungroup()

errado<- ver %>%
  filter(is.na(region_markers)) %>%
  group_by(variable, method, CHR, region_markers) %>%
  summarise("erro" = n)%>%
  ungroup()

acerto<-certo%>%
  count(variable, method, name = "acerto")%>%
  ungroup()

erro<-errado%>%
  group_by(variable, method) %>%
  summarise(erro = sum(erro))
  
coinc <- acerto %>%
  full_join(erro)

coinc <- coinc %>%
  mutate(ngenes = as.integer(
    case_when(
      variable == 1 | variable == 6 ~ 8,
      variable == 2 | variable == 7 ~ 40,
      variable == 3 | variable == 8 ~ 80,
      variable == 4 | variable == 9 ~ 120,
      variable == 5 | variable == 10 ~ 240)
  ),
  porc_acerto = round((acerto/ngenes) * 100, 2),
         porc_erro = round((erro/ngenes) * 100, 2),
  porc_erro=replace(porc_erro, is.na(porc_erro), 0))

coinc <- coinc %>%
  mutate(erro = replace(erro, is.na(erro), 0),
         indice = round(acerto/(acerto+erro)*100,2),
         herd = as.factor(case_when(as.numeric(variable) < 6 ~ "h² = 0.50",
                          as.numeric(variable) > 5 ~ "h² = 0.80")))

library(writexl)
write_xlsx(coinc, "tabela.xlsx")

coinc2<-coinc %>%
  filter(method != "G-BLUP" & method != "MLP" & method != "RBF")

coinc2%>%
  ggplot(aes(as.factor(ngenes), method, fill = porc_acerto)) +
  geom_tile(color = "white") +
  scale_fill_gradient2(
    low = "white",
    high = "red",
    mid = "yellow",
    midpoint = (max(coinc2$porc_acerto) / 2),
    limit = c(0, max(coinc2$porc_acerto)),
    space = "Lab",
    name = "Índice \nde acerto"
  ) +
  facet_wrap(vars(herd),strip.position = "top")+
  theme_minimal() + # minimal theme
  theme(axis.text.x = element_text(
    angle = 45,
    vjust = 1,
    hjust = 1
  ),
  text = element_text(size = 20)) +
  coord_fixed() +
  geom_text(aes(as.factor(ngenes), method, label = porc_acerto),
            color = "black",
            size = 5) +
  theme(
    axis.title.y = element_blank(),
    panel.grid.major = element_blank(),
    panel.border = element_blank(),
    panel.background = element_blank(),
    axis.ticks = element_blank(),
    #legend.justification = c(1, 0),
    legend.position = "top",
    legend.direction = "horizontal",
    legend.title = element_text(size=12),
    legend.text = element_text(size=10)
  ) +
  labs(x = "Número de genes")+
  guides(fill = guide_colorbar(
    barwidth = 7,
    barheight = 1,
    title.hjust = 0.5
  ))

ggsave(
  "percentagem_acerto.tiff",
  width = 12,
  height = 8,
  dpi = 300
)

coinc2%>%
  ggplot(aes(as.factor(ngenes), method, fill = porc_erro)) +
  geom_tile(color = "white") +
  scale_fill_gradient2(
    low = "white",
    high = "red",
    mid = "yellow",
    midpoint = (max(coinc2$porc_erro) / 2),
    limit = c(0, max(coinc2$porc_erro)),
    space = "Lab",
    name = "Índice \nde erro"
  ) +
  facet_wrap(vars(herd),strip.position = "top")+
  theme_minimal() + # minimal theme
  theme(axis.text.x = element_text(
    angle = 45,
    vjust = 1,
    hjust = 1
  ),
  text = element_text(size = 20)) +
  coord_fixed() +
  geom_text(aes(as.factor(ngenes), method, label = porc_erro),
            color = "black",
            size = 5) +
  theme(
    axis.title.y = element_blank(),
    panel.grid.major = element_blank(),
    panel.border = element_blank(),
    panel.background = element_blank(),
    axis.ticks = element_blank(),
    #legend.justification = c(1, 0),
    legend.position = "top",
    legend.direction = "horizontal",
    legend.title = element_text(size=12),
    legend.text = element_text(size=10)
  ) +
  labs(x = "Número de genes")+
  guides(fill = guide_colorbar(
    barwidth = 7,
    barheight = 1,
    title.hjust = 0.5
  ))

ggsave(
  "percentagem_erro.tiff",
  width = 12,
  height = 8,
  dpi = 300
)

coinc2 %>%
ggplot(aes(as.factor(ngenes), method, fill = indice)) +
  geom_tile(color = "white") +
  scale_fill_gradient2(
    low = "white",
    high = "red",
    mid = "yellow",
    midpoint = (max(coinc2$indice) / 2),
    limit = c(0, max(coinc2$indice)),
    space = "Lab",
    name = "Índice \nRelativo"
  ) +
  facet_wrap(vars(herd),strip.position = "top")+
  theme_minimal() + # minimal theme
  theme(axis.text.x = element_text(
    angle = 45,
    vjust = 1,
    hjust = 1
  ),
  text = element_text(size = 20)) +
  coord_fixed() +
  geom_text(aes(as.factor(ngenes), method, label = indice),
            color = "black",
            size = 5) +
  theme(
    axis.title.y = element_blank(),
    panel.grid.major = element_blank(),
    panel.border = element_blank(),
    panel.background = element_blank(),
    axis.ticks = element_blank(),
    #legend.justification = c(1, 0),
    legend.position = "top",
    legend.direction = "horizontal",
    legend.title = element_text(size=12),
    legend.text = element_text(size=10)
  ) +
  labs(x = "Número de genes")+
  guides(fill = guide_colorbar(
    barwidth = 7,
    barheight = 1,
    title.hjust = 0.5
  ))

ggsave(
  "indice_relativo.tiff",
  width = 12,
  height = 8,
  dpi = 300
)


```

